src/aio/search.rs

```diff
@@ -95,81 +107,147 @@ impl SearchFuture {
     }
 
     // Process a control response to extract the control URL
-    fn handle_control_resp(addr: SocketAddr, resp: Bytes) -> Result<String, SearchError> {
+    fn handle_control_resp(addr: SocketAddr, resp: Bytes) -> Result<(String, String), SearchError> {
         debug!("handling control response from: {}", addr);
 
         // Create a cursor over the response data
         let c = std::io::Cursor::new(&resp);
 
         // Parse control URL out of body
-        let url = parsing::parse_control_url(c)?;
+        let urls = parsing::parse_control_urls(c)?;
+
+        Ok(urls)
+    }
+
+    fn request_schemas(
+        addr: SocketAddr,
+        control_schema_url: String,
+    ) -> Result<Pin<Box<dyn Future<Output = Result<Bytes, SearchError>> + Send>>, SearchError> {
+        let client = Client::new();
+        let uri = match format!("http://{}{}", addr, control_schema_url).parse() {
+            Ok(uri) => uri,
+            Err(err) => return Err(SearchError::from(err)),
+        };
+
+        Ok(Box::pin(
+            client
+                .get(uri)
+                .and_then(|resp| hyper::body::to_bytes(resp.into_body()))
+                .map_err(|e| SearchError::from(e)),
+        ))
+    }
 
-        Ok(url)
+    fn handle_schema_resp(resp: Bytes) -> Result<HashMap<String, Vec<String>>, SearchError> {
+        debug!("handling schema response");
+        let c = std::io::Cursor::new(&resp);
+        parsing::parse_schemas(c)
     }
 }
 
 impl Future for SearchFuture {
     type Output = Result<Gateway, SearchError>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<Gateway, SearchError>> {
```

Comment:
If you'd like to turn this into an async function as well, I think it could be much cleaner. You don't have to if you don't feel like it though.

Reply:
Yes, I can rewrite it. As I said, I wanted to preserve much of the code, more so because it was my first time actually working with asynchronous Rust and I didn't know the best practices. But I am positive that I can make it more readable with a nice `async/.await` approach.